modulus/models/gnn_layers/graph.py

```diff
@@ -0,0 +1,365 @@
+        to work on the corresponding local graph according to the partition rank.
+        """
+        if self.is_distributed:
+            return self.dist_graph.get_src_node_features_in_local_graph(local_src_feat)
+        return local_src_feat
+
+    def get_dst_node_features_in_partition(
+        self, global_dst_feat: torch.Tensor
+    ) -> torch.Tensor:
+        """
+        Get local chunk of global destination node features for each rank corresponding
+        to its rank in the process group across which the graph is partitioned.
+        """
+        if self.is_distributed:
+            return self.dist_graph.get_dst_node_features_in_partition(global_dst_feat)
+        return global_dst_feat
+
+    def get_edge_features_in_partition(
+        self, global_efeat: torch.Tensor
+    ) -> torch.Tensor:
+        """
+        Get local chunk of global edge features for each rank corresponding
+        to its rank in the process group across which the graph is partitioned.
+        """
+        if self.is_distributed:
+            return self.dist_graph.get_edge_features_in_partition(global_efeat)
+        return global_efeat
+
+    def get_global_src_node_features(
+        self, local_nfeat: torch.Tensor, get_on_all_ranks: bool = True
+    ) -> torch.Tensor:
+        """
+        Based on local source node features on each rank corresponding
+        to its rank in the process group across which the graph is partitioned,
+        get the global node features either on all group ranks or on group rank 0.
+        """
+        if self.is_distributed:
+            return self.dist_graph.get_global_src_node_features(
+                local_nfeat, get_on_all_ranks
+            )
+        return local_nfeat
+
+    def get_global_dst_node_features(
+        self, local_nfeat: torch.Tensor, get_on_all_ranks: bool = True
+    ) -> torch.Tensor:
+        """
+        Based on local destination node features on each rank corresponding
+        to its rank in the process group across which the graph is partitioned,
+        get the global node features either on all group ranks or on group rank 0.
+        """
+        if self.is_distributed:
+            return self.dist_graph.get_global_dst_node_features(
+                local_nfeat, get_on_all_ranks
+            )
+        return local_nfeat
+
+    def get_global_edge_features(
+        self, local_efeat: torch.Tensor, get_on_all_ranks: bool = True
+    ) -> torch.Tensor:
+        """
+        Based on local edge features on each rank corresponding
+        to its rank in the process group across which the graph is partitioned,
+        get the global edge features either on all group ranks or on group rank 0.
+        """
+        if self.is_distributed:
+            return self.dist_graph.get_global_edge_features(
+                local_efeat, get_on_all_ranks
+            )
+        return local_efeat
+
+    def to(self, *args: Any, **kwargs: Any) -> "CuGraphCSC":
```

Comment:
This is something we have in many places. IIUC, this is about properly handling forward references as we can't use the class itself (here CuGraphCSC) as type when defining the class or methods.

Reply:
Wow, this led me down a fascinating rabbit hole and I learnt something new today! Apparently you can do C++ template style generic types in python and still have annotations and static type checking for it!

WRT this specific comment, I found the documentation for this [here](https://mypy.readthedocs.io/en/stable/generics.html#generic-methods-and-generic-self) and it seems like using `"CuGraphCSC"` would be an explicit annotation but you could also use `TypeVar` so child classes of this can also be typed appropriately. I'll leave it up to you if you want to make this change though it seems somewhat unnecessary here.

Reply:
Ah wow, yeah, the generic type indeed would be better in this case - despite not really necessary.

Reply:
Actually, it is much simpler to just use `typing.Self` as these things indeed just have a `return self`.

Reply:
I'm not sure about using `typing.Self` since it is only supported in python 3.11 onwards according to [PEP 673](https://peps.python.org/pep-0673/). Is there a way to backport it to 3.9/3.10 which is what the NGC pytorch containers use?